utils/logger/circular.go

```diff
@@ -0,0 +1,58 @@
+// pmm-agent
+// Copyright (C) 2018 Percona LLC
+//
+// This program is free software: you can redistribute it and/or modify
+// it under the terms of the GNU Affero General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+//
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU Affero General Public License for more details.
+//
+// You should have received a copy of the GNU Affero General Public License
+// along with this program. If not, see <https://www.gnu.org/licenses/>.
+
+package logger
+
+import (
+	"bufio"
+	"bytes"
+)
+
+type CircularWriter struct {
+	data   []string
+	i      uint32
+	buffer bytes.Buffer
+}
+
+func New(len int) *CircularWriter {
+	writer := CircularWriter{
+		data:   make([]string, len),
+		i:      0,
+		buffer: bytes.Buffer{},
+	}
+	go writer.write()
```

Comment:
`bytes.Buffer` is not thread-safe. It is not possible to write in one goroutine and read in another one without synchronization.